server/src/internalClusterTest/java/org/opensearch/search/SearchWeightedRoutingIT.java

```diff
@@ -158,4 +181,617 @@ public void testSearchWithWRRShardRouting() throws IOException {
+                .prepareSearch("test")
+                .setSize(100)
+                .setQuery(QueryBuilders.matchAllQuery())
+                .execute();
+        }
+
+        logger.info("--> network disruption is stopped");
+        networkDisruption.stopDisrupting();
+
+        for (int i = 0; i < 50; i++) {
+            try {
+                SearchResponse searchResponse = responses[i].get();
+                assertEquals(searchResponse.getFailedShards(), 0);
+                for (int j = 0; j < searchResponse.getHits().getHits().length; j++) {
+                    hitNodes.add(searchResponse.getHits().getAt(j).getShard().getNodeId());
+                }
+            } catch (Exception t) {
+                fail("search should not fail");
+            }
+        }
+
+        assertSearchInAZ("b");
+        assertSearchInAZ("c");
+        assertNoSearchInAZ("a");
+    }
+
+    private void assertNoSearchInAZ(String az) {
+        ImmutableOpenMap<String, DiscoveryNode> dataNodes = internalCluster().clusterService().state().nodes().getDataNodes();
+        String dataNodeId = null;
+
+        for (Iterator<DiscoveryNode> it = dataNodes.valuesIt(); it.hasNext();) {
+            DiscoveryNode node = it.next();
+            if (node.getAttributes().get("zone").equals(az)) {
+                dataNodeId = node.getId();
+                break;
+            }
+        }
+
+        NodesStatsResponse nodeStats = client().admin().cluster().prepareNodesStats().execute().actionGet();
+        for (NodeStats stat : nodeStats.getNodes()) {
+            SearchStats.Stats searchStats = stat.getIndices().getSearch().getTotal();
+            if (stat.getNode().isDataNode()) {
+                if (stat.getNode().getId().equals(dataNodeId)) {
+                    assertEquals(0, searchStats.getQueryCount());
+                    assertEquals(0, searchStats.getFetchCount());
+                }
+            }
+        }
+    }
+
+    private void assertSearchInAZ(String az) {
+        ImmutableOpenMap<String, DiscoveryNode> dataNodes = internalCluster().clusterService().state().nodes().getDataNodes();
+        String dataNodeId = null;
+
+        for (Iterator<DiscoveryNode> it = dataNodes.valuesIt(); it.hasNext();) {
+            DiscoveryNode node = it.next();
+            if (node.getAttributes().get("zone").equals(az)) {
+                dataNodeId = node.getId();
+                break;
+            }
+        }
+
+        NodesStatsResponse nodeStats = client().admin().cluster().prepareNodesStats().execute().actionGet();
+        for (NodeStats stat : nodeStats.getNodes()) {
+            SearchStats.Stats searchStats = stat.getIndices().getSearch().getTotal();
+            if (stat.getNode().isDataNode()) {
+                if (stat.getNode().getId().equals(dataNodeId)) {
+                    Assert.assertTrue(searchStats.getFetchCount() > 0L);
+                    Assert.assertTrue(searchStats.getQueryCount() > 0L);
+                }
```

Comment:
Can we assert for the exact count if possible?

Reply:
Its difficult to get same distribution of shards on every test run, hence we can't assert on actual number. But we have a [task](https://github.com/opensearch-project/OpenSearch/issues/5037) opened to get fail open stats which can help us track this closer.